### DISCUSSION TITLE
Blazor CSLA 7.0 Saving Null value to database when string is Empty.String or " "

### DISCUSSION THREAD

**Discussion by jhuerta3** [MarimerLLC/csla#3518] [General]
Hello, 
Is there a secrete to converting a edit object property value of type string which is Empty.String to null before I save to the database? 
Reason: There is a sql check constraint that will fail with the saved value is Empty.String but when its null its okay. 
I have added nullable value for the property type on the Edit object and the Info object but when it saves to the database its still Empty.String and not null. 

        public static readonly PropertyInfo<string?> XXXProperty = RegisterProperty<string?>(nameof(XXX));
        [Display(Name = "XXX")]
        public string? XXX
        {
            get { return GetProperty(XXXProperty); }
            set { SetProperty(XXXProperty, value); }
        }

**Comment by rockfordlhotka** [comment]
The CSLA property implementation for `string` values does have code specifically to ensure that `string.Empty` is the default value. This is because some data binding implementations (WinForms and/or WPF iirc) don't work with null string values.

Does your business logic need to differentiate between "" and null? Or is that just some artifact of a database designer making an arbitrary choice?

Usually a db uses null to differentiate between a value that was never entered by the user, vs "" which is a valid empty value entered by the user. Not everyone _gets_ that this is the purpose of null, so there's a lot of weird stuff out there.

**Comment by hurcane** [comment]
We manage this discrepancy between UI vs. database by using a DBValue extension method within our data portal methods. We enforce usage through code reviews, but we have considered a custom analysis rule to detect direct assignment of strings to database parameters instead of using the DBValue method.

**Comment by jhuerta3** [comment]
Hello @rockfordlhotka  and @hurcane,

Basically, the SQL check constraints does distinguish between null and "".  If the value is null for the column, then the check is bypassed but it has a value "" or anything else the check is verified and "" fails the check. Basically, I wanted to see if I could get away from doing something like this on the Business Edit object Update method.   
                  
tReTransmitEngID = string.IsNullOrEmpty(tReTransmitEngID) ? null : tReTransmitEngID, 

or this on the Dal object 

old.tReTransmitEngID = string.IsNullOrEmpty(data.tReTransmitEngID) ? null : data.tReTransmitEngID;
I was just wondering if I can set some value the business edit object  tReTransmitEngID property that would convert "" to null. 
I thoughtthis would it " string? " not no. 

        public static readonly PropertyInfo<string?> tReTransmitEngIDProperty = RegisterProperty<string?>(nameof(tReTransmitEngID));
        [Display(Name = "tReTransmitEngID")]
        public string? tReTransmitEngID
        {
            get { return GetProperty(tReTransmitEngIDProperty); }
            set { SetProperty(tReTransmitEngIDProperty, value); }
        }

The both of the above sample lines work. I just implemented the one on the business edit update method. 
Thank you.
